File: SWEA/D3/11315. 오목 판정/오목 판정.py
from collections import deque
T = int(input())
dx = [0, 1, 1, 1]
dy = [1, 0, 1, -1]

for tc in range(1, T + 1):
    N = int(input())
    arr = [list(input()) for _ in range(N)]
    result = 'NO'
    for i in range(N):
        for j in range(N):
            if arr[i][j] == 'o':
                for k in range(4):
                    x,y,cnt = i,j,0
                    while 0<= x < N and 0<= y < N and arr[x][y] == 'o':
                        cnt += 1
                        x += dx[k]
                        y += dy[k]
                    if cnt >= 5:
                        result = 'YES'

    print(f'#{tc} {result}')

File: SWEA/D3/11315. 오목 판정/오목 판정.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Solution {
	
    public static void main(String[] args) throws IOException {
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	
    	int t = Integer.parseInt(br.readLine());
    	int[] dx = {1,0,1,1};
    	int[] dy = {0,1,1,-1};
    	
    	for (int test = 1; test <= t; test++) {
    		int n = Integer.parseInt(br.readLine());
    		char[][] arr = new char[n][n];
    		
    		for (int i = 0; i < n; i++) {
    			String s = br.readLine();
    			for (int j = 0; j < n; j++) {
					arr[i][j] = s.charAt(j);
				}
			}
    		
    		String ans = "NO";
    		
    		for (int i = 0; i < n; i++) {
				for (int j = 0; j < n; j++) {
					
					if(arr[i][j] == 'o') {
						for (int d = 0; d < 4; d++) {
							int cnt = 0;
							int nx = i;
							int ny = j;
							
							while(0 <= nx && nx < n && 0 <= ny && ny < n && arr[nx][ny] == 'o') {
								cnt++;
								nx += dx[d];
								ny += dy[d];
							}
							
							if(cnt >= 5) {
								ans = "YES";
							}
						}
					}
				}
			
			}
    		System.out.printf("#%d %s\n",test,ans);   				
		}
    }
}
